packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt

```diff
@@ -0,0 +1,236 @@
+        return _realmChanged.asSharedFlow()
+            .map {
+                Pair(it.dbPointer, VersionId(RealmInterop.realm_get_version_id(it.dbPointer)))
+            }.distinctUntilChanged { old, new ->
+                old.second == new.second
+            }
+//        return callbackFlow {
+//            val token: AtomicRef<Cancellable> = kotlinx.atomicfu.atomic(NO_OP_NOTIFICATION_TOKEN)
+//            withContext(dispatcher) {
+//                token.value = addRealmChangedListener { frozenRealm ->
+//                    val result = trySend(frozenRealm)
+//                    if (!result.isSuccess) {
+//                        // FIXME What to do if this fails?
+//                        throw IllegalStateException("Notification could not be handled: $result")
+//                    }
+//                }
+//            }
+//            awaitClose {
+//                token.value.cancel()
+//            }
+//        }
+    }
+
+    fun addRealmChangedListener(callback: Callback<Pair<NativePointer, VersionId>>): Cancellable {
+        TODO("Waiting for RealmInterop to have support for global Realm changed")
+    }
+
+    /**
+     * Listen to changes to a RealmResults
+     */
+    fun <T : RealmObject> resultsChanged(results: RealmResults<T>): Flow<RealmResults<T>> {
+        return callbackFlow {
+            val token: AtomicRef<Cancellable> = kotlinx.atomicfu.atomic(NO_OP_NOTIFICATION_TOKEN)
+            withContext(dispatcher) {
+                ensureActive()
+                val newToken = addResultsChangedListener(results) { frozenResults ->
+                    // Realm should already have been updated with the latest version
+                    // So `owner` should as a minimum be at the same version as the notification Realm.
+                    val result = trySend(frozenResults)
+                    checkResult(result)
+                }
+                token.value = newToken
+            }
+            awaitClose {
+                token.value.cancel()
+            }
+        }
+    }
+
+    // Verify that notifications emitted to Streams are handled in an uniform manner
+    private fun checkResult(result: ChannelResult<Unit>) {
+        if (result.isClosed) {
+            // If the Flow was closed, we assume it is on purpose, so avoid raising an exception.
+            return
+        }
+        if (!result.isSuccess) {
+            // TODO Is there a better way to handle this?
+            throw IllegalStateException("Notification could not be sent: $result")
+        }
+    }
+
+    // FIXME Need to expose change details to the user
+    //  https://github.com/realm/realm-kotlin/issues/115
+    /**
+     * Register a change listener on a live RealmResults. All objects returned in the callback are frozen.
+     */
+    fun <T : RealmObject> addResultsChangedListener(results: RealmResults<T>, callback: Callback<RealmResults<T>>): Cancellable {
```

Comment:
This must be called on the notifier thread? Otherwise we wouldn't be able to thaw the result and actually register for notifications!? Maybe add a comment and make it internal.

Reply:
Yes, I added a comment in the Dokka about it. So far this is only being used by our Flows